dfu: mcuboot: add shell module

mniestroj opened the pull request:
Add new shell module for mcuboot enabled application. It provides
information about image slots and allows to perform such operations as:
confirm, erase and request upgrade.

Example output after requesting an upgrade:
```
uart:~$ mcuboot
swap type: test
confirmed: 1

primary area (1):
  version: 0.10.0+0
  image size: 368660
  magic: good
  swap type: none
  copy done: unset
  image ok: set

secondary area (2):
  version: 12.13.14+0
  image size: 368660
  magic: good
  swap type: test
  copy done: unset
  image ok: unset
```

Example output test booting new image:
```
uart:~$ mcuboot
swap type: revert
confirmed: 0

primary area (1):
  version: 12.13.14+0
  image size: 368660
  magic: good
  swap type: test
  copy done: set
  image ok: unset

secondary area (2):
  version: 0.10.0+0
  image size: 368660
  magic: unset
  swap type: none
  copy done: unset
  image ok: unset
```

mniestroj committed:
dfu: mcuboot: add shell module

Add new shell module for mcuboot enabled application. It provides
information about image slots and allows to perform such operations as:
confirm, erase and request upgrade.

Signed-off-by: Marcin Niestroj <m.niestroj@grinn-global.com>

utzig commented:
The implementation is fine, but the erase command is a bit weird:

`$ mcuboot erase 0`: this works corretly but erases de bootloader
`$ mcuboot erase 1`: this hangs by erasing the running image (if not using some XIP mode!)

Should we allow those things to happen?

nvlsianpu commented on subsys/dfu/boot/mcuboot_shell.c:
```diff
@@ -0,0 +1,200 @@
+
+	id = strtoul(argv[1], NULL, 0);
+
+	err = boot_erase_img_bank(id);
```
exacts area_id is defined in compile time,
User might erase any area which match given `ID`, not only these `IDs` which belongs to image banks.

nvlsianpu replied:
This command should limit `ID` used somehow.

mniestroj replied:
I wonder if we should limit `ID` in `mcuboot_shell.c` implementation or rather inside `boot_erase_img_bank()` body. In general I share your concerns here, about (not) giving user possibility to do stupid stuff over shell. However I have the same concern when directly using `boot_erase_img_bank()`, which is needed when something other than mcumgr is supposed to be used for DFU. I've written `mcuboot_shell.c` to basically get developers the same access as using C API directly, but in a much easier way, hence using shell interface.

So in summary I lean towards adding a check in `boot_erase_img_bank()`.

@nvlsianpu @utzig please share your opinion o that.

utzig replied:
> So in summary I lean towards adding a check in `boot_erase_img_bank()`.

Yes, this would be right approach IMO.

mniestroj replied:
I've created PR #30060, where we can discuss how to restrict `boot_erase_img_bank()`. I will leave this PR as is.

nvlsianpu commented:
Just one minor request, otherwise this PR looks fine.
Thanks!

mniestroj commented:
I have rebased on master branch and added a check on `FLASH_AREA_IMAGE_SECONDARY` which is now optionally defined.

I think that we can move the discussion of restricting erase area id to #30060, which doesn't look that trivial.

mniestroj commented:
@nvlsianpu @utzig Is there something you want me to do in this PR? I think we agreed to check permission of erase in `boot_erase_img_bank()`.

nvlsianpu commented on subsys/dfu/boot/mcuboot_shell.c:
```diff
@@ -0,0 +1,202 @@
+	shell_print(shell, "  image size: %u",
+		    (unsigned int) hdr.h.v1.image_size);
+
+	err = boot_read_swap_state_by_id(area->id, &swap_state);
```
boot_read_swap_state_by_id() was never before considered as API function. It was non static just because migt be build form mcuboot or zephyr surcecode. Value of `struct boot_swap_state` is used to determine the swap-status by int `boot_swap_type_multi(int image_index)` and boot_swap_type().
In #30370 i swich dfu/boot library to use mcuboot surce code - so `boot_swap_type_multi(int image_index)` doesn't present in the API. Of course it might, but personally I belive that boot_swap_type fucntions are more suitable here as user is rather interested in current swap type, and not how to determine it form boot_swap_status structure fields values.

nvlsianpu replied:
I mean to use int mcuboot_swap_type(void) and I can introduced `int mcuboot_swap_type_multi(int image_index)`.

nvlsianpu replied:
@mniestroj what do you think about my concerns?

mniestroj replied:
I've used `struct boot_swap_state` for debugging firmware upgrade process when having some problems, such as https://github.com/mcu-tools/mcuboot/issues/829. I agree that in "normal cases" user should not need such information, which is even kind of confusing when user is unfamiliar with mcubot internals.

Keep in mind that `mcuboot_shell.c` already prints information from `mcuboot_swap_type()`. So the question now is whether we can remove this per-slot information about image trailers:
```
  magic: unset
  swap type: none
  copy done: unset
  image ok: unset
```

I would leave that information if easily available (which is exactly why I have that exposed in this PR, as we had non-static functions already). If however functions currently used to fetch this information are not intended to be public (from mcuboot perspective), then let's remove above information. I don't have strong opinion here.

nvlsianpu replied:
I decided to try current command  behavior https://github.com/zephyrproject-rtos/zephyr/pull/30370, see `int mcuboot_read_swap_state_by_id()`.
